orangecontrib/geo/widgets/owmap.py

```diff
@@ -1,906 +1,849 @@
-        north, east, south, west, width, height, zoom, origin, map_pane_pos = self._drawing_args
-
-        lat, lon = self._latlon_data.T
-        visible = ((lat <= north) & (lat >= south) &
-                   (lon <= east) & (lon >= west)).nonzero()[0]
-        in_subset = (np.in1d(self.data.ids, self._subset_ids)
-                     if self._subset_ids.size else
-                     np.tile(True, len(lon)))
-
-        is_js_path = self.is_js_path = len(visible) < self.N_POINTS_PER_ITER
-
-        self._update_legend(is_js_path)
+        future = self.loader.get(t)
+        @future.add_done_callback
+        def set_tile(_future):
+            if _future.cancelled():
+                return
 
-        np.random.shuffle(visible)
-        # Sort points in subset to be painted last
-        visible = visible[np.lexsort((in_subset[visible],))]
+            assert _future.done()
 
-        if is_js_path:
-            self.evalJS('clear_markers_overlay_image()')
-            self._update_js_markers(visible, in_subset[visible])
-            self._owwidget.disable_some_controls(False)
-            return
+            if _future.exception():
+                _t = _future.exception().tile
+                _t.n_loadings += 1
+                # retry to download image
+                self._load_one_from_net(_t)
+            else:
+                _t, img = _future.result()
+                if not _t.disc_cache:
+                    self._show_internet_error(False)
+                self.mem_cache[_t] = img
+                self._add_patch(_t)
+                self._update_map_item()
+
+        self.futures.append(future)
+
+    def _show_internet_error(self, show):
+        if not self.master.Warning.no_internet.is_shown() and show:
+            self.master.Warning.no_internet()
+        elif self.master.Warning.no_internet.is_shown() and not show:
+            self.master.Warning.no_internet.clear()
+
+    def _cancel_futures(self):
+        for future in self.futures:
+            future.cancel()
+            if future._reply is not None:
+                future._reply.close()
+                future._reply.deleteLater()
+                future._reply = None
+
+    def set_tile_provider(self, tp):
+        self.tile_provider = tp
+        self.view_box.set_tile_provider(self.tile_provider)
+        self.tile_attribution.setHtml(self.tile_provider.attribution)
+
+    def clear_map(self):
+        self._cancel_futures()
+        self.futures = []
+        self.map = None
+        self.tz = 1
+
+        self.plot_widget.removeItem(self.b_map_item)
+        self.b_map_item = self.__new_map_item(-3)
+        self.plot_widget.addItem(self.b_map_item)
+
+        self.plot_widget.removeItem(self.map_item)
+        self.map_item = self.__new_map_item(-2)
+        self.plot_widget.addItem(self.map_item)
 
-        self.evalJS('clear_markers_js();')
-        self._owwidget.disable_some_controls(True)
+    def clear(self):
+        if self.freeze:
```

Comment:
It is usually better to call parents method despite overriden. What if a new property is added to the parent? Therefore never cleared.
So is this case, plot can be fully cleared and the map reploted (if the zooming properties can be easily remembered, of course).